### PR TITLE
ENH: Add np.scale function

### PR DESCRIPTION
**Pull request by kernc**
This PR adds a scale function for scaling non-NaN array values in any dimension onto a desired interval. I found myself somewhat regularly using something like this and copying it into projects. I think a robust implementation should fit into numpy quite nicely.

I hope you too won't find it unimaginable to have something like this in numpy and at roughly the proposed location.

### CODE REVIEW

File: numpy/lib/function_base.py [hunk 1]
```diff
@@ -5010,3 +5011,76 @@ def append(arr, values, axis=None):
...
+    Returns
+    -------
+    scaled_array : ndarray
+        Output array with the same shape as `arr` and all non-NaN values
+        scaled to interval [min, max].
+
+    Raises
+    ------
+    ValueError
+        When ``min > max`` or when values of ``min``, ``max``, and ``axis``
```
Comment (eric-wieser) [inline numpy/lib/function_base.py]: `min > max` seems like an unnecessary restriction

File: numpy/lib/function_base.py [hunk 2]
```diff
@@ -5010,3 +5011,76 @@ def append(arr, values, axis=None):
...
+    if out is None:
+        out = np.empty_like(arr, dtype=float)
+
+    mins, maxs = np.nanmin(arr, axis), np.nanmax(arr, axis)
+
+    if axis is None:
+        out[...] = (arr - mins) / (maxs - mins) * (max - min) + min
+    else:
+        arr = np.rollaxis(arr, axis)
+        out[...] = np.rollaxis(
```
Comment (eric-wieser) [inline numpy/lib/function_base.py]: I feel like this could be done better by just broadcasting with `min` and `max`

File: numpy/lib/tests/test_function_base.py [hunk 3]
```diff
@@ -1,4 +1,4 @@
-from __future__ import division, absolute_import, print_function
+from __future__ import division, absolute_import, print_function, with_statement
```
Comment (eric-wieser) [inline numpy/lib/tests/test_function_base.py]: Pretty sure numpy doesn't need 2.5 support any more

File: THANKS.txt [hunk 4]
```diff
@@ -49,6 +49,7 @@ Alan McIntyre for updating the NumPy test framework to use nose, improve
 Joe Harrington for administering the 2008 Documentation Sprint.
 Mark Wiebe for the new NumPy iterator, the float16 data type, improved
     low-level data type operations, and other NumPy core improvements.
+Kernc for scale.
```
Comment (eric-wieser) [inline THANKS.txt]: Is there a threshold for being added to this list? What is that threshold?

Comment (rgommers) [inline THANKS.txt]: There's not, the contributor guidelines suggest you add yourself. Full name would be better than GitHub handle.

We should probably change that file in some way or get rid of it completely. We acknowledge and thank all contributors in the release notes; this file is often not updated; when you do update it in a PR there's often a merge conflict.

**Comment by rgommers** [comment]
@kernc can you please propose adding this function on the numpy-discussion mailing list? That's where we decide on enhancements.

The name ``scale`` is a bit too generic imho.

File: numpy/lib/function_base.py (hunk 1 above)
Comment (kernc) [inline numpy/lib/function_base.py]: Right. I was thinking about that but then apparently forgot about it. :smile:

File: numpy/lib/function_base.py (hunk 2 above)
Comment (kernc) [inline numpy/lib/function_base.py]: Thanks. I agree it's much better. Have I gotten the code correctly and concise enough?

File: numpy/lib/function_base.py [hunk 5]
```diff
@@ -5010,3 +5011,77 @@ def append(arr, values, axis=None):
...
+            raise ValueError('Invalid shape of min / max. Shapes should be: {}'
+                             .format(valid_shapes[-1]))
+
+    if out is None:
+        out = np.empty_like(arr, dtype=float)
+
+    mins, maxs = np.nanmin(arr, axis), np.nanmax(arr, axis)
+
+    if axis is not None:
+        shape = [slice(None)] * arr.ndim
```
Comment (eric-wieser) [inline numpy/lib/function_base.py]: shape is a bad name here. `slice` or `idx` would be better

File: numpy/lib/function_base.py [hunk 6]
```diff
@@ -5010,3 +5011,77 @@ def append(arr, values, axis=None):
...
+    if axis is not None:
+        valid_shapes = ((), arr.shape[:axis] + arr.shape[axis + 1:])
+        if min.shape not in valid_shapes or max.shape not in valid_shapes:
+            raise ValueError('Invalid shape of min / max. Shapes should be: {}'
+                             .format(valid_shapes[-1]))
+
+    if out is None:
+        out = np.empty_like(arr, dtype=float)
+
+    mins, maxs = np.nanmin(arr, axis), np.nanmax(arr, axis)
```
Comment (eric-wieser) [inline numpy/lib/function_base.py]: Add `keepdims=True` here, and you don't need the `mins, maxs = mins[shape], maxs[shape]` line below

File: numpy/lib/function_base.py [hunk 7]
```diff
@@ -5010,3 +5011,77 @@ def append(arr, values, axis=None):
...
+    >>> np.scale(a, [0, 1, 0], [1, 1, 1], axis=0)
+    array([[ 0.,  1.,  0.],
+           [ 1.,  1.,  1.]])
+    """
+    arr, min, max = asanyarray(arr), asanyarray(min), asanyarray(max)
+
+    if axis is None and (min.ndim or max.ndim):
+        raise ValueError('min and max must be scalar when axis is None')
+    if axis is not None:
+        valid_shapes = ((), arr.shape[:axis] + arr.shape[axis + 1:])
```
Comment (eric-wieser) [inline numpy/lib/function_base.py]: Would it make more sense to enforce that the shape is ` arr.shape[:axis] + (1,) + arr.shape[axis + 1:]`?

Comment (kernc) [inline numpy/lib/function_base.py]: I was considering it, but then anyone would have to do this hard-to-reason-about reshaping themselves?

Comment (kernc) [inline numpy/lib/function_base.py]: I mean, in a simple 2d case, it seems easier to obtain a 1d vector of extremes than a properly shaped matrix ...

On the other hand, `keepdims` looks ubiquitous enough. I think you're right. :+1:

Comment (eric-wieser) [inline numpy/lib/function_base.py]: Perhaps both could be supported?

**Comment by kernc** [comment]
> The name scale is a bit too generic imho.

The transformation where something is linearly changed in size is normally [called scaling](https://www.willamette.edu/~gorr/classes/GeneralGraphics/Transforms/transforms2d.htm#Scaling). People [search](https://stackoverflow.com/questions/5294955/how-to-scale-down-a-range-of-numbers-with-a-known-min-and-max-value) [for](https://stats.stackexchange.com/questions/70801/how-to-normalize-data-to-0-1-range) [it](https://gamedev.stackexchange.com/questions/45358/scale-a-normalized-2d-vector-always-to-the-same-length). With vectors, it's also called normalization, but the meaning of this term seems more domain dependent (what is normal?) and implies scaling to unit length or to 1, whereas this function is meant for scaling to arbitrary intervals given interval extremes. What would you name it?

I'll write on the mailing list, thanks.

**Comment by shoyer** [comment]
I'm still not sure this is warranted for numpy, but for the name, I think `rescale` would be more appropriate.

**Comment by eric-wieser** [comment]
I feel like the function that's actually wanted here is the `remap` provided in arduino, which this function can be written in terms of:
```python
def scale(data, min, max, axis):
    return np.remap(data,
        in_min=data.min(axis, keepdims=True),
        in_max=data.max(axis, keepdims=True),
        out_min=min,
        out_max=max
    )
```

Of course, `remap` can probably be expressed in terms of `interp`, but I think that might be worth providing in the library

**Comment by charris** [comment]
1.12.x has been branched, you should move the comments in the 1.12.0-notes to the 1.13.0-notes.

**Comment by mattharrigan** [comment]
FWIW, I use [this ](http://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.scale.html#sklearn.preprocessing.scale) function pretty regularly.  That does a different operation,  by default returning an array with zero mean and a standard deviation of 1.  That seems like a different use case.  Not sure if this function can or should include the sklearn functionality.  In either event the name scale to me means the sklearn version.

**Comment by kernc** [comment]
@mattharrigan [Both methods](https://en.wikipedia.org/wiki/Feature_scaling#Methods) are valid, mostly depending on the use case.

I like @eric-wieser's idea about additional prior min / prior max arguments. Will update accordingly. Two use cases I experienced in the past week: 

1. Have a long vector of data. Because it's really long, work with samples. To visualize the sample, associate color intensity (opacity, 0-255) to each value according to the peak-to-peak of the original full vector.
2. Use `scipy.optimize.differential_evolution`. Provide varied, non-normalized bounds for _N_ variables. Provide a loss function that works within those bounds, and consider including some (i.e. L1) [regularization](https://en.wikipedia.org/wiki/Regularization_%28mathematics%29) of parameters. Because not all parameters are from the same interval space (i.e. some are bounded by _[0, 1]_, others by _[-1e5, 1e5]_), I want to scale the parameters according to input bounds so that all parameters exert the same regularization influence. I could pre-scale the bounds, of course, but that might need making the loss function less intuitive.

@shoyer, `rescale` it is. How strongly do you feel against this in numpy and why?

**Comment by shoyer** [comment]
> How strongly do you feel against this in numpy and why?

I'm not strongly opposed, just leaning against it. Consider this a [-0 vote](http://apache.org/foundation/voting.html#expressing-votes-1-0-1-and-fractions) (Note that you shouldn't take my opinion as a definitive pronouncement here, but you will need to at least a plurality of maintainers that this is worth adding.)

My reasons:
1. It's not unambiguous what `scale` or even `rescale` means. The existence of `sklearn.preprocessing.scale` doing something totally different is actually a strong point *against* including this in NumPy. NumPy already has too many utility functions with domain specific logic that we are stuck maintaining for perpetuity (for a recent example, see [atleast_3d](https://github.com/numpy/numpy/pull/7804#issuecomment-256968781)).
2. This is a small utility function, which is quite easy to write (and rewrite) in user or domain specific packages (e.g., `sklearn.preprocessing.scale`). So the incremental value of including it in NumPy is minimal.

**Comment by charris** [comment]
I feel pretty much the same as @shoyer about this. Long term, my concern is that numpy may slowly become bloated with small functions. That said, the numpy polynomials do this sort of scaling but they also need to store the scaling parameters. That points up the problem of providing a single function for this simple operation when the use cases may differ in small details from project to project.

**Comment by eric-wieser** [comment]
>the numpy polynomials do this sort of scaling 

Could you elaborate with an example?

**Comment by charris** [comment]
@eric-wieser The various polynomial classes scale the domain to the window. For instance, the default window for Chebyshev polynomials is [-1, 1], while the domain can be any interval. This sort of scaling is essential for numerical stability, especially when fitting polynomials to data.

**Comment by eric-wieser** [comment]
Oh, I see, by "do this sort of scaling", you mean "use it internally", not "already implement the feature of this PR"

**Comment by charris** [comment]
Yep. Scaling and shifting of some sort is quite common, but I think the guestions here are

- Is scaling complicated enough to justify a function in NumPy.
- Can we make a function of sufficient generality to cover most use cases.

Note that scaling might not only be scaling to intervals, but scaling by standard deviation or some other norm.

**Comment by eric-wieser** [comment]
Pushing the `rescale(array, in_min, in_max, out_min=0, out_max=1)` I suggested earlier, rescaling to the uniform normal is expressed as:

    np.rescale(arr, mu, mu+std)

**Comment by mherkazandjian** [comment]
I have my own implementation of this function and I use it often. It would be nice to have it in numpy.
But to avoid having lots of small function in numpy, i think it would be nice to have a flexible design for this "rescale" functionality. A design that would allow for providing a custom scaling function, i.e the scaling could be uniform, normal..etc..

btw, there are similar implementations of this:
  http://www.harrisgeospatial.com/docs/BYTSCL.html

File: numpy/lib/function_base.py [hunk 8]
```diff
@@ -5010,3 +5011,77 @@ def append(arr, values, axis=None):
...
+        if min.shape not in valid_shapes or max.shape not in valid_shapes:
+            raise ValueError('Invalid shape of min / max. Shapes should be: {}'
+                             .format(valid_shapes[-1]))
+
+    if out is None:
+        out = np.empty_like(arr, dtype=float)
```
Comment (bashtage) [inline numpy/lib/function_base.py]: You could move the test for out = None to the return so that you wouldn't create an innecessary out matrix in the most likely case when out is `None`.

```
temp = ... # Compute
if out is not None:
    out[...] = temp # Copy to out
    return out
return temp
```

Comment (kernc) [reply numpy/lib/function_base.py]: Thanks. 👍

**Comment by WarrenWeckesser** [comment]
This pull request has had the "needs decision" tag since November 2017.  My impression is that the function is at the edge of what is considered worthwhile adding to NumPy.  So this comment is really a *ping* to the NumPy devs: is it desirable to add such a linear rescaling function to NumPy?  The lack of a decision is effectively a "no", but maybe a few devs taking a fresh look will think otherwise.

@kernc, if the answer turns out to be "yes", are you still interested in working on this?

If the answer to both those questions is yes, there is still some work to do, starting with a rebase to fix the merge conflict.

The function should do a better job of handling a constant input.  Currently it generates several warnings and returns all `nan`s:

```
In [2]: np.rescale([10, 10, 10])                                                                                                   
/Users/warren/mc37numpy/lib/python3.7/site-packages/numpy-1.15.0.dev0+967809c-py3.7-macosx-10.7-x86_64.egg/numpy/lib/function_base.py:4479: RuntimeWarning: divide by zero encountered in double_scalars
  offset = (in_max * out_min - in_min * out_max) / oldlen
/Users/warren/mc37numpy/lib/python3.7/site-packages/numpy-1.15.0.dev0+967809c-py3.7-macosx-10.7-x86_64.egg/numpy/lib/function_base.py:4480: RuntimeWarning: divide by zero encountered in true_divide
  scale = newlen / oldlen
/Users/warren/mc37numpy/lib/python3.7/site-packages/numpy-1.15.0.dev0+967809c-py3.7-macosx-10.7-x86_64.egg/numpy/lib/function_base.py:4482: RuntimeWarning: invalid value encountered in add
  res = arr * scale + offset
Out[2]: array([nan, nan, nan])
```
